### PR TITLE
Add support for complex filters

### PR DESCRIPTION
**Pull request by sophiegreen**

### CODE REVIEW

**Comment by xiaoping8385** [comment]
Exception: hudson.AbortException: Fail: the returned code 404 is not in the accepted range: [[100‥399]]

**Comment by stewartfrancis** [comment]
Whoops, forgot to leave a comment on the review.  Looks really good!  Just had a few questions, and there seemed to be a TODO we might want to look into.  That escaping's all customizeable now

File: tests/unit/modules/test_cmci_filters.py [hunk 1]
```diff
@@ -0,0 +1,402 @@
...
+                    }
+                ]
+            }
+        }
+    })
+
+
+def test_complex_filter_operator(cmci_module):  # type: (CMCITestHelper) -> None
+    records = [{'name': 'bat', 'dsname': 'STEWF.BLOP.BLIP'}]
+    # TODO currently this is wrong as it adds %C2 before the operator, which CPSM doesn't like.
```
[the remark this answers is not in the document]
Comment (sophiegreen) [reply tests/unit/modules/test_cmci_filters.py:141]: %C2 is Â. It's only doing it for the !=, the rest are fine. Yeah that's after your changes, which fixed all the other bits (thank you!), I was going to take a look at how you fixed those next to see if this is something we can fix.

File: plugins/module_utils/cmci.py [hunk 2]
```diff
@@ -405,3 +468,80 @@ def append_attributes(element, attributes):
...
+            'operator': operator_dict,
+            'value': value_dict,
+            **my_dict
+        },
+        'required_together': [('attribute', 'value')]
+    }
+
+
+def _convert_filter_operator(operator):
+    if operator is None: return '='
```
[the remark this answers is not in the document]
Comment (sophiegreen) [reply plugins/module_utils/cmci.py:501]: Good point!

File: plugins/module_utils/cmci.py [hunk 3]
```diff
@@ -405,3 +468,80 @@ def append_attributes(element, attributes):
...
+    if operator == '<' or operator == 'LT': return '<'
```
[the remark this answers is not in the document]
Comment (sophiegreen) [reply plugins/module_utils/cmci.py:502]: ah thanks, hadn't seen that format 👍

File: plugins/module_utils/cmci.py [hunk 4]
```diff
@@ -243,11 +258,47 @@ def get_resources_request_params(self):  # type: () -> Dict[str, str]
...
+            if complex_filter:
+                complex_filter_string = ''
+                if not request_params:
+                    request_params = {}
+
+                and_item = complex_filter['and']
+                or_item = complex_filter['or']
+                attribute_item = complex_filter['attribute']
+
+                if and_item is not None and or_item is not None:
```
[the remark this answers is not in the document]
Comment (sophiegreen) [reply plugins/module_utils/cmci.py:282]: not sure what you mean?

**Comment by xiaoping8385** [comment]
Exception: hudson.AbortException: Fail: the returned code 404 is not in the accepted range: [[100‥399]]

File: plugins/module_utils/cmci.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (stewartfrancis) [reply plugins/module_utils/cmci.py:283]: You're checking that `and` and `or` are mutually exclusive.  Is `attribute` not mutually exclusive with them too?

File: plugins/module_utils/cmci.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (stewartfrancis) [reply plugins/module_utils/cmci.py]: tbh you could just do
```python
if operator == 'LT':
  return '<'
...
else:
  return operator
```

File: tests/unit/modules/test_cmci_filters.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (stewartfrancis) [reply tests/unit/modules/test_cmci_filters.py:141]: I wonder if that's an encoding problem.  Create an issue for it and we'll look at it this afternoon

[the remark this answers is not in the document]
Comment (sophiegreen) [reply tests/unit/modules/test_cmci_filters.py:141]: Issue #17 raised

File: plugins/module_utils/cmci.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (sophiegreen) [reply plugins/module_utils/cmci.py:283]: Good question, I'm going to see if this works with CMCI....

[the remark this answers is not in the document]
Comment (sophiegreen) [reply plugins/module_utils/cmci.py:283]: ok yep, this works fine with both attribute and 'and' at the same level.
Tested by writing a new unit test for it and using the output on 56:  http:///cicsex56.hursley.ibm.com:28953/CICSSystemManagement/cicslocalfile/CICSEX56/IYCWEMW2?CRITERIA=%28FILE%3D%27DFHCSD%27%29%20AND%20%28FILE%3D%27DFHCSD%27%29%20AND%20%28FILE%3D%27DFHCSD%27%29

**Comment by xiaoping8385** [comment]
Exception: hudson.AbortException: Fail: the returned code 404 is not in the accepted range: [[100‥399]]

**Comment by xiaoping8385** [comment]
Exception: hudson.AbortException: Fail: the returned code 404 is not in the accepted range: [[100‥399]]

**Review by stewartfrancis — APPROVED** [review]

**Comment by xiaoping8385** [comment]
Exception: hudson.AbortException: Fail: the returned code 404 is not in the accepted range: [[100‥399]]

**Pull request merged**